### PR TITLE
Support socket() call with proto==0

### LINKED ISSUE #18873: zsock_socket() should support proto==0
**Is your enhancement proposal related to a problem? Please describe.**
Most implementations of BSD-style sockets, both those aiming at comprehensive POSIX compliance (e.g. Linux, macOS, *BSD) and those intentionally deviating from it (e.g. WinSock, lwIP), allow calling `socket()` with the `proto` argument set to 0, which causes the protocol to be defaulted to a typical choice.

E.g. `socket(AF_INET, SOCK_STREAM, 0)` is equivalent to `socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)` and `socket(AF_INET, SOCK_DGRAM, 0)` is equivalent to `socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)`.

This is actually mandated by POSIX (see [here](http://pubs.opengroup.org/onlinepubs/9699919799/functions/socket.html) and [here](http://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_10_18)), but as I mentioned above, even implementations that do not aim for full POSIX compliance typically support this usage.

This makes it harder to port some libraries and applications that would otherwise be trivially portable to Zephyr, but use `socket()` in this way.

**Describe the solution you'd like**
I'd like `zsock_socket()` to behave as POSIX-compliant `socket()` would for this case. A minimal viable solution would be to make (AF_INET|AF_INET6, SOCK_STREAM) default the protocol to `IPPROTO_TCP` and (AF_INET|AF_INET6, SOCK_DGRAM) default the protocol to `IPPROTO_UDP`, as these are the two most common cases.

### PR DESCRIPTION
**Pull request by jukkar**
Fixes #18873

### CODE REVIEW

**Comment by pfalcon** [comment]
> Support socket() call with proto==0

It's very weird that wasn't done before. I actually would test a hypothesis that it was there, but was lost.

**Comment by pfalcon** [comment]
> It's very weird that wasn't done before. I actually would test a hypothesis that it was there, but was lost.

I grepped git history, and yeah, it was never done. Weird.

File: subsys/net/lib/sockets/sockets.c [hunk 1]
```diff
@@ -131,6 +131,16 @@ int zsock_socket_internal(int family, int type, int proto)
 
 int z_impl_zsock_socket(int family, int type, int proto)
 {
+	if (proto == 0) {
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: Nack. This should be done, per address-family, in an address family implementation handler.

**Review by pfalcon — CHANGES_REQUESTED** [review]
Thanks for coming forward to fix this, but lets follow the architecture of socket type handlers which yourself introduced. A handler for a particular socket family/type known what's the default protocol for it. There's no need to add adhoc checks for that to the outside code. And that doesn't scale to the arbitrary socket families/types anyway.

File: subsys/net/lib/sockets/sockets.c (hunk 1 above)
Comment (jukkar) [reply subsys/net/lib/sockets/sockets.c]: This support has nothing to do with address family. Only thing important here is the type and proto so no need to duplicate the code in individual socket handler.

Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c]: OMG, this is something. Do you try to research fixes you make, or just sprinkle some random code around?

`man socket`:

>       The protocol specifies a  particular  protocol  to  be  used  with  the
>       socket.  Normally only a single protocol exists to support a particular
>       socket type within a given protocol family, in which case protocol  can
>       be  specified  as  0.   However, it is possible that many protocols may
>       exist, in which case a particular protocol must be  specified  in  this
       manner. 

So, no need to spread handling of socket() parameters around the codebase. The code which implements socket() (and that's socket handler) knows best how to handle the params.

Comment (jukkar) [reply subsys/net/lib/sockets/sockets.c]: > The code which implements socket() (and that's socket handler) knows best how to handle the params.

Yep, that is why the check is in socket().
See also https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_10_18
We can certainly add here checks for `family==AF_INET || family == AF_INET6 || family == AF_PACKET`

Comment (jukkar) [reply subsys/net/lib/sockets/sockets.c]: > OMG, this is something. Do you try to research fixes you make, or just sprinkle some random code around?

BTW, there is no need to write these "witty" remarks in your comments, it just makes harder for you to get your message through. If you find a bug in the commit, just say so.

**Comment by jukkar** [comment]
Added checks for the family.

File: subsys/net/lib/sockets/sockets.c (hunk 1 above)
Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c]: > BTW, there is no need to write these "witty" remarks

I'm sorry, but there's a stream of a patches from you (great work!), and I have a concern that a solution in almost each of them wasn't researched thoroughly enough. We're at the stage where a) we have a complex system; b) it mostly works. Any careless changes may throw it back into havoc. And if not, just make harder to maintain and debug (it already is!)

Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c]: > Yep, that is why the check is in socket().

socket() is nothing but a dispatcher which call whatever socket handler ready to handle the input params: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/sockets/sockets.c#L132 . We don't do any param checks there, it makes no sense - there can be arbitrary socket families, types, protocols, and only a handler for a particular combo (such handlers being configurable!) knows how to handle those params.

> See also https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_10_18

And what we read there? "When a socket is created in the **Internet family** with a protocol value of zero ..."

Once again, next day, there can be AF_KABOOM, and only socket handler for that family may know what it means to have protocol == 0 for it.

**Comment by pfalcon** [comment]
> Added checks for the family.

No, that's not what I call for. Instead:

1. There should be no patching of z_impl_zsock_socket().
2. Handling of protocol == 0 for AF_INET, AF_INET6 shoul dbe added to zsock_socket_internal() - that's our default socket handler, which handles AF_INET, AF_INET6.
3. When time comes, handling of protocol == 0 for other sock families/types will be added to corresponding socket module. E.g., AF_CAN will be handled in sockets_can.c, and not somewhere else.

**Comment by jukkar** [comment]
That is one option although it means duplicating the check for packet socket as it is handled separately.

**Comment by pfalcon** [comment]
> That is one option although it means duplicating the check for packet socket as it is handled separately.

But not duplicate, but performing work (here: checks) related to a particular socket handler in that socket handler, and not somewhere else. That was the reason why you introduced independent socket handlers in there first place, isn't it? Do encapsulate code related to a particular socket kind, and decouple it from the rest of code. For example, we could disable AF_INET* sockets and leave other sockets, and that would work without extra #ifdef'ing the checks somewhere outside, etc.

That's actually what caught me attention here: we define a flexible and advanced infrastructure, they seemingly forget about it and apply random patchings elsewhere instead of fully benefiting from that infra.

**Comment by jukkar** [comment]
> No, that's not what I call for. Instead:
> 
>     1. There should be no patching of z_impl_zsock_socket().

There isn't any more

> 
>     2. Handling of protocol == 0 for AF_INET, AF_INET6 shoul dbe added to zsock_socket_internal() - that's our default socket handler, which handles AF_INET, AF_INET6.

It is there now.

@pfalcon Can you re-check?

**Review by pfalcon — APPROVED** [review]
Looks good, thanks!

**Pull request merged**